doc/designs/quic-design/server/quic-server-api.md

```diff
@@ -0,0 +1,521 @@
+support multiple outgoing client connections on one socket these will need to
+support a non-zero-length local CID. Equally servers will need to support CIDs.
+
+By default when a QLSO is used, we will use a real, non-zero-length CID. If the
+existing QUIC client API is used (with no QLSO) we will likely spin up a
+`QUIC_PORT` internally but specially choose to use a zero-length local CID since
+we know this port will never be used for more than one connection.
+
+If an application creates a QLSO explicitly but knows it will not need to
+create more than one connection, it can pass the following flag when creating
+the listener:
+
+```c
+#define SSL_LISTENER_FLAG_SINGLE_CONN   (1U << 1)
+```
+
+Failure to do so is nonfatal and simply increases packet overheads.
+
+### Future Compatibility: TLS over TCP Support
+
+Assuming we choose to use the SSL object type as the basis of a new listener
+object type, let us consider how this API could also be used to support TLS over
+TCP, at least in a server role:
+
+```c
+/* Blocking example: */
+{
+    SSL *l, *conn;
+    BIO *tcp_listener = get_bio_sock_for_listening_tcp_socket();
+    SSL_CTX *ctx = create_ssl_ctx(TLS_server_method());
+    BIO_ADDR addr;
+
+    l = SSL_new_listener(ctx);
+    SSL_set_bio(l, tcp_listener, tcp_listener);
+
+    for (;;) {
+        conn = SSL_accept_conn(l, 0); /* standard SSL_CONNECTION is returned */
+
+        spawn_thread_to_process_conn(conn, &addr);
+    }
+}
+```
+
+Essentially, when `SSL_new_listener` is given a `SSL_CTX` which is using a TLS
+method, it returns some different kind of object, say a `TCP_LISTENER`, which
+also is a kind of SSL object. `SSL_accept_conn` calls `accept(2)` via some BIO
+abstraction on the provided BIO, and constructs a new BIO on the resulting
+socket, wrapping it via standard `SSL_new` using the same `SSL_CTX`.
+
+Thus, this API should be compatible with any future adaptation to also support
+TCP.
+
+### Future Compatibility: DTLS over UDP Support
+
+The relevant concerns regarding DTLS are likely to be highly similar to QUIC,
+so this API should also be adaptable to support DTLS in the future, providing
+a unified server-side API. `SSL_new_listener` will return e.g. a
+`DTLS_LISTENER`.
+
+### Supporting Multiple Sockets
+
+There are three means that we can provide to support a server which wants to
+listen on multiple UDP sockets:
+
+1. Allow a QLSO to bind multiple BIOs;
+2. Create multiple QLSOs and use some polling mechanism to service all of them;
+3. Create some higher level object which can aggregate multiple QLSOs.
+
+(1) would require more extensive API changes and should only be done if a
+compelling reason to do so arises. (3) is a more advanced design which is
+theoretically possible in future and is discussed in the next section.
+For now, we will support only (2) (which we can support via our planned polling
+API) until a compelling need arises.
```

Comment:
Yeah. The basic idea is once we need to do this to support migration/multipath, we introduce the QDSO and have multiple QLSOs under it.

The alternative is that we come up with some new API to bind multiple BIOs to a single SSL object, which seems a bit awkward (it would require us to assign some kind of unique identifier for each BIO attached so an application can ask us to unregister it later, etc.). You would still end up with the same object hierarchy internally as well.

In any case, this is for the future, so we're not making a commitment here for QUIC R2.